marshal/claim.go

```diff
@@ -85,6 +86,21 @@ func newClaim(topic string, partID int, marshal *Marshaler, consumer *Consumer,
 	log.Debugf("[%s:%d] consumer offsets: early = %d, cur/comm = %d/%d, late = %d",
 		topic, partID, offsets.Earliest, offsets.Current, offsets.Committed, offsets.Latest)
 
+	// For offsets, we strictly prefer the contents of the MarshalTopic and will use that
+	// if present. If we don't have that data, then we'll fall back to the Kafka committed
+	// offsets. Failing that we'll start at the beginning of the partition.
+	if offsets.Current > 0 {
+		// Ideal case, we just use the Marshal offset that is already set
+	} else if offsets.Committed > 0 {
+		log.Infof("[%s:%d] no Marshal offset found, using committed offset %d",
+			topic, partID, offsets.Committed)
+		offsets.Current = offsets.Committed
+	} else {
+		log.Infof("[%s:%d] no Marshal or committed offset found, using earliest offset %d",
+			topic, partID, offsets.Earliest)
+		offsets.Current = offsets.Earliest
+	}
+
 	// Take the greatest of the committed/current offset, this means we will recover from
```

Comment:
If we have a rewind operation in which the `CommitOffset` fails but the `Heartbeat` succeeds, could it be the case that the rewind gets "undone" by the block below (the block that starts with `if offsets.Committed > 0 && offsets.Committed > offsets.Current`)?